### PR TITLE
[TFLite] Fix FuseUnpackAndConcatToReshape incorrectly fusing permutations as reshape

### LINKED ISSUE #116047: The TFLite converter incorrectly eliminates a non-trivial data permutation and replaces it with an identity mapping.
```
import io
import sys

try:
    import numpy as np
    import tensorflow as tf
    import torch
    import onnxruntime as ort
except ImportError as e:
    print(f"missing dep: {e}", file=sys.stderr)
    sys.exit(2)

# ── Test input: shape=(4, 3), values 1..12 ─────────────────────────────────
X = np.array(
    [[1., 2., 3.],
     [4., 5., 6.],
     [7., 8., 9.],
     [10., 11., 12.]],
    dtype=np.float32,
)
H, W = X.shape

# Reference (numpy): trace the op chain explicitly
def numpy_ref(x):
    a = x.reshape(H, W, 1)
    b = np.concatenate(list(np.split(a, W, axis=1)), axis=0)
    return b.reshape(H, W)

EXPECTED = numpy_ref(X)
print(f"input    : {X.flatten().tolist()}")
print(f"expected : {EXPECTED.flatten().tolist()}")
print()

results = []  # (runtime, output, ok)

# ── TF Eager (Keras) ────────────────────────────────────────────────────────
class M(tf.keras.Model):
    def __init__(self):
        super(M, self).__init__()
    @tf.function(input_signature=[tf.TensorSpec([H, W], tf.float32)])
    def call(self, x):
        a = tf.reshape(x, [H, W, 1])
        b = tf.concat(tf.unstack(a, axis=1), 0)
        return tf.reshape(b, [H, W])

m = M()
keras_out = m(tf.constant(X)).numpy()
results.append(("Keras eager", keras_out, np.allclose(keras_out, EXPECTED)))

# ── XLA (tf.function jit_compile=True) ─────────────────────────────────────
@tf.function(jit_compile=True, input_signature=[tf.TensorSpec([H, W], tf.float32)])
def f_xla(x):
    a = tf.reshape(x, [H, W, 1])
    b = tf.concat(tf.unstack(a, axis=1), 0)
    return tf.reshape(b, [H, W])

xla_out = f_xla(tf.constant(X)).numpy()
results.append(("XLA", xla_out, np.allclose(xla_out, EXPECTED)))

# ── PyTorch eager ───────────────────────────────────────────────────────────
xt = torch.tensor(X)
a = xt.reshape(H, W, 1)
b = torch.cat(list(torch.unbind(a, dim=1)), dim=0)
torch_out = b.reshape(H, W).numpy()
results.append(("PyTorch", torch_out, np.allclose(torch_out, EXPECTED)))

class TM(torch.nn.Module):
    def forward(self, x):
        a = x.reshape(H, W, 1)
        b = torch.cat(list(torch.unbind(a, dim=1)), dim=0)
        return b.reshape(H, W)

buf = io.BytesIO()
torch.onnx.export(TM(), (xt,), buf, input_names=["x"], output_names=["y"], opset_version=17)
sess = ort.InferenceSession(buf.getvalue(), providers=["CPUExecutionProvider"])
ort_out = sess.run(None, {"x": X})[0]
results.append(("ONNX Runtime", ort_out, np.allclose(ort_out, EXPECTED)))

tfl = tf.lite.TFLiteConverter.from_keras_model(m).convert()
itp = tf.lite.Interpreter(
    model_content=tfl,
    experimental_op_resolver_type=tf.lite.experimental.OpResolverType.BUILTIN_WITHOUT_DEFAULT_DELEGATES,
)
in_idx = itp.get_input_details()[0]["index"]
itp.resize_tensor_input(in_idx, [H, W])
itp.allocate_tensors()
itp.set_tensor(in_idx, X)
itp.invoke()
tflite_out = itp.get_tensor(itp.get_output_details()[0]["index"])
tflite_ok = np.allclose(tflite_out, EXPECTED)
results.append(("TFLite", tflite_out, tflite_ok))

# ── Report ──────────────────────────────────────────────────────────────────
print(f"{'Runtime':<14} {'Output':<60} {'Match?'}")
print("-" * 86)
for name, out, ok in results:
    status = "OK" if ok else "WRONG"
    print(f"{name:<14} {str(out.flatten().tolist()):<60} {status}")

# Exit code: 0 if TFLite reproduces the bug, 1 otherwise
sys.exit(0 if not tflite_ok else 1)
```

  - Test case: shape=(4, 3), values 1..12
  - Expected: [1, 4, 7, 10, 2, 5, 8, 11, 3, 6, 9, 12]
  - TFLite output: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12] (input unchanged)
  - All other runtimes (Keras eager, XLA, PyTorch, ONNX Runtime): correct

### PR DESCRIPTION
**Pull request by ssam18**
The `FuseUnpackAndConcatToReshape` optimization was blindly replacing any `unpack + concat` sequence with a `reshape`, but this is only correct when the concat axis matches the unpack axis. When they differ, the two ops together perform a data permutation that a plain reshape cannot replicate.

Added a guard that bails out of the fusion when the normalized concat axis differs from the unpack axis, and added a regression test covering the mismatched-axis case. Fixes #116047

### CODE REVIEW

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thanks for the contribution!
  
  Upon verifying the changes, we found a bug related to type conversion that breaks the existing test suite when negative axes are used. 
  
  ### Issue Details:
  `concat_op.getAxis()` and `unpack_op.getAxis()` return a 32-bit representation. When assigning them directly to `int64_t` variables:
  ```cpp
  int64_t unpack_axis = unpack_op.getAxis();
  int64_t concat_axis = concat_op.getAxis();
  
  ```
it results in a zero-extension rather than a sign-extension. Consequently, a negative axis like -1 becomes 4294967295, which causes the normalization comparison to fail and disables the optimization.

**Suggested Fix:**
Please cast the getters to int32_t first to ensure correct sign-extension when represented as int64_t:

```cpp
  int64_t unpack_axis = static_cast<int32_t>(unpack_op.getAxis());
  ...
  int64_t concat_axis = static_cast<int32_t>(concat_op.getAxis());
```

**Unit Tests:**
The regression test you included (noFuseUnpackAndConcatToReshapeWhenAxesDiffer in optimize.mlir) is excellent and necessary to safeguard against invalid fusions. No additional unit tests are required, but please apply the casting fix above so that the existing tests (like @fuseUnpackAndConcatToReshape which uses axis = -1) will pass successfully.

**Comment by ssam18** [comment]
> Thanks for the contribution!
> 
> Upon verifying the changes, we found a bug related to type conversion that breaks the existing test suite when negative axes are used.
> 
> ### Issue Details:
> `concat_op.getAxis()` and `unpack_op.getAxis()` return a 32-bit representation. When assigning them directly to `int64_t` variables:
> 
> ```c++
> int64_t unpack_axis = unpack_op.getAxis();
> int64_t concat_axis = concat_op.getAxis();
> ```
> 
> it results in a zero-extension rather than a sign-extension. Consequently, a negative axis like -1 becomes 4294967295, which causes the normalization comparison to fail and disables the optimization.
> 
> **Suggested Fix:** Please cast the getters to int32_t first to ensure correct sign-extension when represented as int64_t:
> 
> ```c++
>   int64_t unpack_axis = static_cast<int32_t>(unpack_op.getAxis());
>   ...
>   int64_t concat_axis = static_cast<int32_t>(concat_op.getAxis());
> ```
> 
> **Unit Tests:** The regression test you included (noFuseUnpackAndConcatToReshapeWhenAxesDiffer in optimize.mlir) is excellent and necessary to safeguard against invalid fusions. No additional unit tests are required, but please apply the casting fix above so that the existing tests (like @fuseUnpackAndConcatToReshape which uses axis = -1) will pass successfully.

Thanks @dmiltr3, good catch on the zero-extension. Fixed by casting both getAxis() results to int32_t before normalization, so -1 keeps its sign. The existing fuseUnpackAndConcatToReshape (axis = -1) test passes now.

**Review by dmiltr3 — APPROVED** [review]

**Review by dmiltr3 — APPROVED** [review]

**Review by dmiltr3 — APPROVED** [review]

**Review by dmiltr3 — APPROVED** [review]
